File: backend/models/DeletionRequest.js
const mongoose = require('mongoose');

const deletionRequestSchema = new mongoose.Schema({
  targetType: {
    type: String,
    enum: ['project', 'payment', 'milestone'],
    required: true,
  },
  targetId: { type: mongoose.Schema.Types.ObjectId, required: true },
  targetName: { type: String, required: true },
  requestedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  requestedByName: { type: String, default: '' },
  approvals: {
    developer: { type: Boolean, default: null },
    boss: { type: Boolean, default: null },
    accountant: { type: Boolean, default: null },
    developerUid: { type: String, default: null },
    bossUid: { type: String, default: null },
    accountantUid: { type: String, default: null },
    developerAt: { type: Date, default: null },
    bossAt: { type: Date, default: null },
    accountantAt: { type: Date, default: null },
  },
  status: {
    type: String,
    enum: ['pending', 'partiallyApproved', 'approved', 'rejected'],
    default: 'pending',
  },
  createdAt: { type: Date, default: Date.now },
});

deletionRequestSchema.methods.toPublicJSON = function () {
  return {
    id: this._id.toString(),
    targetType: this.targetType,
    targetId: this.targetId.toString(),
    targetName: this.targetName,
    requestedBy: this.requestedBy.toString(),
    requestedByName: this.requestedByName,
    approvals: {
      developer: this.approvals.developer,
      boss: this.approvals.boss,
      accountant: this.approvals.accountant,
      developerUid: this.approvals.developerUid,
      bossUid: this.approvals.bossUid,
      accountantUid: this.approvals.accountantUid,
      developerAt: this.approvals.developerAt ? this.approvals.developerAt.toISOString() : null,
      bossAt: this.approvals.bossAt ? this.approvals.bossAt.toISOString() : null,
      accountantAt: this.approvals.accountantAt ? this.approvals.accountantAt.toISOString() : null,
    },
    status: this.status,
    createdAt: this.createdAt.toISOString(),
  };
};

module.exports = mongoose.model('DeletionRequest', deletionRequestSchema);

File: lib/models/deletion_request.dart
import 'payment.dart';

enum DeletionTargetType { project, payment, milestone }

class DeletionRequest {
  final String id;
  final DeletionTargetType targetType;
  final String targetId;
  final String targetName;
  final String requestedBy;
  final String requestedByName;
  final PaymentApproval approvals;
  final PaymentStatus status;
  final DateTime createdAt;

  DeletionRequest({
    required this.id,
    required this.targetType,
    required this.targetId,
    required this.targetName,
    required this.requestedBy,
    this.requestedByName = '',
    required this.approvals,
    required this.status,
    required this.createdAt,
  });

  factory DeletionRequest.fromMap(Map<dynamic, dynamic> map) {
    final approvals = PaymentApproval.fromMap(
      (map['approvals'] as Map<dynamic, dynamic>?) ?? {},
    );
    return DeletionRequest(
      id: map['id'] ?? '',
      targetType: map['targetType'] == 'project'
          ? DeletionTargetType.project
          : map['targetType'] == 'milestone'
              ? DeletionTargetType.milestone
              : DeletionTargetType.payment,
      targetId: map['targetId'] ?? '',
      targetName: map['targetName'] ?? '',
      requestedBy: map['requestedBy'] ?? '',
      requestedByName: map['requestedByName'] ?? '',
      approvals: approvals,
      status: _computeStatus(approvals),
      createdAt:
          DateTime.tryParse(map['createdAt']?.toString() ?? '') ?? DateTime.now(),
    );
  }

  static PaymentStatus _computeStatus(PaymentApproval approvals) {
    if (approvals.isRejected) return PaymentStatus.rejected;
    if (approvals.isFullyApproved) return PaymentStatus.approved;
    if (approvals.hasAnyApproval) return PaymentStatus.partiallyApproved;
    return PaymentStatus.pending;
  }

  String get statusDisplayName {
    switch (status) {
      case PaymentStatus.pending:
        return 'Pending Approval';
      case PaymentStatus.partiallyApproved:
        return 'Partially Approved';
      case PaymentStatus.approved:
        return 'Approved (Deleted)';
      case PaymentStatus.rejected:
        return 'Rejected';
    }
  }

  String get targetTypeDisplayName {
    switch (targetType) {
      case DeletionTargetType.project:
        return 'Project';
      case DeletionTargetType.payment:
        return 'Payment';
      case DeletionTargetType.milestone:
        return 'Milestone';
    }
  }
}
